youtube_dl/extractor/youku.py

```diff
@@ -1,123 +1,172 @@
+                b[h], b[f] = b[f], b[h]
+                h += 1
+            q = f = h = 0
+            while q < len(c):
+                h = (h + 1) % 256
+                f = (f + b[h]) % 256
+                b[h], b[f] = b[f], b[h]
+                if isinstance(c[q], int):
+                    result += chr(c[q] ^ b[(b[h] + b[f]) % 256])
+                else:
+                    result += chr(ord(c[q]) ^ b[(b[h] + b[f]) % 256])
+                q += 1
+
+            return result
+
+        e_code = trans_e(f_code_1, base64.b64decode(bytes(ep, 'ascii')))
+        sid, token = e_code.split('_')
+        new_ep = trans_e(f_code_2, '%s_%s_%s' % (sid, vid, token))
+        return base64.b64encode(bytes(new_ep, 'latin')), sid, token
+
+    def parse_m3u8(self, cm):
+        raw_urls = re.findall(r'(https?:.+?)\?', cm)
+        t_url = raw_urls[0]
+        urls = []
+        urls.append(t_url)
+        for url in raw_urls:
+            if url != t_url:
+                urls.append(url)
+                t_url = url
+        return urls
+
+    def parse_ext_l(self, fm, supported_format):
+        if fm in ('hd3', 'hd2', 'flvhd', 'flv'):
+            ext = 'flv'
+        elif fm in ('mp4',):
+            ext = 'mp4'
+        elif fm[:3] == '3gp':
+            ext = '3gp'
+        return ext
 
     def _real_extract(self, url):
         mobj = re.match(self._VALID_URL, url)
         video_id = mobj.group('id')
 
-        info_url = 'http://v.youku.com/player/getPlayList/VideoIDS/' + video_id
+        json_url = 'http://v.youku.com/player/getPlayList/VideoIDS/%s/Pf/4/ctype/12/ev/1' % video_id
 
-        config = self._download_json(info_url, video_id)
+        w_info = self._download_json(json_url, video_id)
+        data = w_info['data'][0]
 
-        error_code = config['data'][0].get('error_code')
+        error_code = data.get('error_code')
         if error_code:
             # -8 means blocked outside China.
-            error = config['data'][0].get('error')  # Chinese and English, separated by newline.
-            raise ExtractorError(error or 'Server reported error %i' % error_code,
-                                 expected=True)
+            # Chinese and English, separated by newline.
+            error = data.get('error')
+            raise ExtractorError(
+                error or 'Server reported error %i' %
+                error_code,
+                expected=True)
 
-        video_title = config['data'][0]['title']
-        seed = config['data'][0]['seed']
+        title = data['title']
+        #seed = data['seed']
 
         format = self._downloader.params.get('format', None)
-        supported_format = list(config['data'][0]['streamfileids'].keys())
-
-        # TODO proper format selection
-        if format is None or format == 'best':
-            if 'hd2' in supported_format:
-                format = 'hd2'
+        supported_format = data['streamtypes']
+
+        # DONE proper format selection
```

Comment:
Remove this comment, the only purpose of the `TODO` comment was to remind that it had to be implemented.